### PR TITLE
feat: create recoverable keys from dashboard

### LINKED ISSUE #2097: Allow setting `recoverable: true` when creating keys in the dashboard
### Preliminary Checks

- [X] I have reviewed https://unkey.com/docs for existing features that would solve my problem

- [X] I have searched for existing feature requests: https://github.com/unkeyed/unkey/issues

- [X] This issue is not a question, general help request, or anything other than a feature request directly related to Unkey. Please ask questions in our Discord community: https://unkey.com/discord.

### Is your feature request related to a problem? Please describe.

> if you guys end up implementing the ability to create recoverable keys from dashboard that'd be awesome tho. rn we're iterating quickly and auth is evolving a lot so much faster to work with dashboard to generate test keys than have a million different server functions for generating keys.
>
> also same thing for ownerId vs externalId. rn if I create a key in the dashboard it doesn't add an externalId and ownerId is marked as deprecated all over api docs. so if I try to query based on only externalId any dashboard created keys won't show up bc they're using ownerId
>
> would be great to switch dashboard over to externalId if thats the way forward or at least provide the option to add/modify externalId so api and dashboard work seamlessly together.

### Describe the solution

Add a toggle to mark a key as recoverable in the createKey UI
CleanShot 2024-09-17 at 09 14 44@2x

The trpc handler needs to be updated as well, the logic can be copied from https://github.com/unkeyed/unkey/blob/36d43366d2c4fad74264c5e9234fceeed1e4afa4/apps/api/src/routes/v1_keys_createKey.ts#L367

### Describe alternatives you have considered (if any)

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by harshsbhat**
## What does this PR do?

Fixes #2097

This PR allows users to create recoverable keys through the dashboard, which is only possible through API as of now.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

- Test A: 

Make sure you have store_encrypted_keys enabled. Head over to the dashboard and create a new key http://localhost:3000/apis/api_id/keys/ks_id/new with `recoverable: true`.  Get the key_id of the newly created key and use it with `getKey` API like this ( make sure you have appropriate permissions to get the decrypted key ):

```
curl --request GET \
  --url 'http://localhost:8787/v1/keys.getKey?keyId=key_id123&decrypt=true' \
  --header 'Authorization: Bearer unkey_rootKey'
```
  
The decrypted key should be logged in the response
  
- Test B (few edge cases that might be worth testing ):

1. Without `store_encrypted_keys` the recoverable card is not visible. 
2. Added in the transaction so even if the encryption fails an extra key won't be added.

## Checklist

https://www.loom.com/share/713ebadf169b43eb9055dda724dd7008?sid=82985e82-072f-4a10-9595-78c010d7484f

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [Contributing Guide](./CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand areas
- [x] Ran `pnpm build`
- [x] Ran `pnpm fmt`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues

### Appreciated

- [x] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the Unkey Docs if changes were necessary

### CODE REVIEW

**Comment by harshsbhat** [comment]
third try 😔

hopefully, I get it this time

File: apps/dashboard/lib/trpc/routers/key/create.ts [hunk 1]
```diff
@@ -110,7 +113,38 @@
...
+          };
+          const requestId = crypto.randomUUID();
+          const context: RequestContext = { requestId };
+          const vaultRes = await vault.encrypt(context, encryptReq).catch((err) => {
+            console.error(err);
+            throw new TRPCError({
+              code: "INTERNAL_SERVER_ERROR",
+              message: "Encryption Failed. Please contact support using support@unkey.dev",
+            });
+          });
```
[the remark this answers is not in the document]
Comment (harshsbhat) [reply apps/dashboard/lib/trpc/routers/key/create.ts:130]: This is fine:
-    keyring: workspace?.id,
+    keyring: workspace.id,

I doubt the other two change requests tho. Let's wait for some maintainer

**Comment by harshsbhat** [comment]
It is building the dashboard correctly on the local setup.

Not sure why it is failing here. Can you send me the fail message if it is on my side?

**Comment by perkinsjr** [comment]
Yup it's failing for the `@unkey/vault` type import

```
@unkey/dashboard:build: Type error: Cannot find module '@unkey/vault' or its corresponding type declarations.
@unkey/dashboard:build: 
@unkey/dashboard:build:    5 | import { newId } from "@unkey/id";
@unkey/dashboard:build:    6 | import { newKey } from "@unkey/keys";
@unkey/dashboard:build: >  7 | import { type EncryptRequest, type RequestContext, Vault } from "@unkey/vault";
@unkey/dashboard:build:      |                                                                 ^
@unkey/dashboard:build:    8 | import { z } from "zod";
@unkey/dashboard:build:    9 | import { auth, t } from "../../trpc";
@unkey/dashboard:build:   10 |
@unkey/dashboard:build:  ELIFECYCLE  Command failed with exit code 1.
@unkey/dashboard:build: ERROR: command finished with error: command (/vercel/path0/apps/dashboard) /pnpm8/node_modules/.bin/pnpm run build exited (1)
@unkey/dashboard#build: command (/vercel/path0/apps/dashboard) /pnpm8/node_modules/.bin/pnpm run build exited (1)
  Tasks:    4 successful, 5 total
```

**Comment by harshsbhat** [comment]
Hey @perkinsjr, I added it in the package.json and also pushed the pnpm-lock. Can you check it now?

**Comment by harshsbhat** [comment]
Hey can you take a look at this? I think it is building correctly now. @perkinsjr

**Comment by harshsbhat** [comment]
@perkinsjr @chronark 

Is there any issue with this PR? If you've decided to handle this internally, that's perfectly fine; we can close it.

**Comment by perkinsjr** [comment]
> @perkinsjr @chronark
> 
> Is there any issue with this PR? If you've decided to handle this internally, that's perfectly fine; we can close it.

We haven't handle this internally we just have a lot on our plate currently. We will prioritize this next week

**Comment by perkinsjr** [comment]
If recoverable keys are enabled on your account, you can now create a key in the dashboard and it will be recoverable by default.
